src/libsystemd-network/ndisc-option.c

```diff
@@ -853,3 +1169,94 @@ int ndisc_option_get_mac(Set *options, uint8_t type, struct ether_addr *ret) {
                 *ret = p->mac;
         return 0;
 }
+
+int ndisc_send(int fd, const struct sockaddr_in6 *dst, const struct icmp6_hdr *hdr, Set *options) {
+        int r;
+
+        assert(fd >= 0);
+        assert(dst);
+        assert(hdr);
+
+        struct iovec *iov = NULL;
+        size_t n_iov = 0;
+        CLEANUP_ARRAY(iov, n_iov, iovec_array_free);
+
+        iov = new(struct iovec, 1 + set_size(options));
+        if (!iov)
+                return -ENOMEM;
+
+        r = ndisc_header_size(hdr->icmp6_type);
+        if (r < 0)
+                return r;
+        size_t hdr_size = r;
+
+        _cleanup_free_ uint8_t *copy = newdup(uint8_t, hdr, hdr_size);
+        if (!copy)
+                return -ENOMEM;
+
+        iov[n_iov++] = IOVEC_MAKE(TAKE_PTR(copy), hdr_size);
+
+        const sd_ndisc_option *option;
+        SET_FOREACH(option, options) {
```

Comment:
Since this function goes over all the options and their fields it would be great to add it to the ndisc fuzz target. I think now that `ndisc_parse_options` can parse packets directly it should be possible to drop the kludgy pipe along with its artificial 2048-byte limit and just stuff bytes into packets, parse them and then pass the options on to `ndisc_send`. It should help to exercise those code paths with all sorts of options. MSan should be happy too:
```c
/* SPDX-License-Identifier: LGPL-2.1-or-later */

#include "fd-util.h"
#include "fuzz.h"
#include "icmp6-packet.c"
#include "ndisc-option.h"

int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
        _cleanup_(icmp6_packet_unrefp) ICMP6Packet *p = NULL;
        _cleanup_set_free_ Set *options = NULL;
        _cleanup_close_ int fd = -EBADF;
        static const struct sockaddr_in6 dst = {
                .sin6_family = AF_INET6,
                .sin6_addr = IN6ADDR_ALL_ROUTERS_MULTICAST_INIT,
        };
        int r;

        if (outside_size_range(size, 0, 65535))
                return 0;

        fuzz_setup_logging();

        p = icmp6_packet_new(size);
        if (!p)
                return 0;

        memcpy(p->raw_packet, data, size);

        r = icmp6_packet_verify(p);
        if (r < 0)
                return 0;

        r = ndisc_parse_options(p, &options);
        if (r < 0)
                return 0;

        fd = open("/dev/null", O_RDWR|O_CLOEXEC|O_NOCTTY);
        assert_se(fd >= 0);

        (void) ndisc_send(fd, &dst, icmp6_packet_get_header(p), options);

        return 0;
}
```

Reply:
Thanks. A similar test case is added to fuzz-ndisc-rs.